Add details to Prometheus Exporter requirement (#2124)
open-telemetry/opentelemetry-specification: specification/metrics/sdk_exporters/prometheus.md

**Status**: [Experimental](../../document-status.md)

> Prometheus Exporter is a [MetricExporter](../sdk.md#metricexporter) which reacts
> to the Prometheus scraper and reports the metrics passively to
> [Prometheus](https://prometheus.io/), or [pushes
> metrics](https://prometheus.io/docs/instrumenting/pushing/) to a [Prometheus
> Pushgateway](https://github.com/prometheus/pushgateway).

Reviewer: - The "Prometheus Exporter" noun phrase needs an indefinite article.
- Remove mention of the push functionality give we are not including it and it should not be a part of the definition.
- Clarify the request/response relationship with word choice.
- Clarify the data format being communicate (metric aggregate values, right?).

```suggestion
A Prometheus Exporter is a [MetricExporter](../sdk.md#metricexporter) which responds
to [Prometheus](https://prometheus.io/) scraper requests with collected metric aggregate
values.
```

Author: If we put a strong wording which indicates Prometheus Exporter is a "Pull Exporter" (e.g. responds to scraper...), later (after initial Stable release) when we add push support it'll be a breaking change (so we have to either bump the major version of the spec, or we need to introduce another exporter, e.g. `PrometheusPushExporter`).

Reviewer: > ... or we need to introduce another exporter, e.g. `PrometheusPushExporter`

I would support this.

Reviewer 2: FWIW, Pull and Push are also separate Prometheus exporters in Otel Collector: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter
The Push one is called prometheusremotewrite exporter.

Author: I've updated the PR based on the discussions and comments. PTAL.

**Status**: [Experimental](../../document-status.md)

Prometheus Exporter is a [MetricExporter](../sdk.md#metricexporter) which reacts
to the Prometheus scraper and reports the metrics passively to
[Prometheus](https://prometheus.io/), or [pushes
metrics](https://prometheus.io/docs/instrumenting/pushing/) to a [Prometheus
Pushgateway](https://github.com/prometheus/pushgateway).

Prometheus Exporter SHOULD support [Pull mode](../sdk.md#pull-metric-exporter),

> if possible.

Reviewer: It seems like this should be strengthened to a requirement. It is the default and expected operation of any prometheus system and would be a compatibility issue if one implementation does not support this.

Author: I think this is already the strongest wording we can put here, considering the fact that some languages/runtimes don't allow HTTP server (e.g. JavaScript running inside a browser, embedded C module running on IoT devices).

Author: I've updated the PR. Regarding this comment I think we have several options:
1. Make the pull mode a `MUST`, which means JavaScript running inside a browser probably won't be able to support it (and will leave `N/A` in the feature matrix).
2. Use `SHOULD` + "if possible".

Reviewer: Since there isn't a requirement to implement these exporter, and this is written with the expected audience of implementations that can implement this I would prefer to go with option (1). That way it would be clearly defined that if an implementation supports a Prometheus exporter they MUST support pull based metrics. Having implementation that cannot support this report an `N/A` seems appropriate.

Reviewer: Saying "SHOULD if possible" is a poor way to specify a normative statement. If we want to go this route it would be better to describe the cases where it is not possible explicitly. Otherwise, this opens up the specification to ambiguity of subjective definition of what "if possible" is.

Author: I've updated the PR based on the discussions and comments. PTAL.

Prometheus Exporter MAY support [Push mode](../sdk.md#push-metric-exporter), if
there is a need to support short-lived applications (e.g. short-lived
service-level batch jobs, JavaScript applications running inside web browsers,

> etc.).

Reviewer 3: Here's the prometheus remote write [specification](https://docs.google.com/document/d/1LPhVRSFkGNSuU1fBd81ulhsCPR4hkSZyyBj1SZ8fWOM/edit#). While remote write is certainly useful for all the reasons that push metrics are generally useful, the doc has an interesting blurb:

> The remote write protocol is not intended for use by applications to push metrics to Prometheus Remote Write compatible receivers.  It is intended that a Prometheus remote write compatible sender scrapes instrumented applications or exporters and sends remote write messages to a server.

Maybe prometheus has changed its tune and it more inclined to encourage applications to use remote write; however as originally conceived, it seems that was not the intent. 

I think we should leave push out of the initial feature set, and potentially come back to it post release if there is demand.

Reviewer 4: I think we should remove ths paragraph and not mention push-prometheus because it comes with a lot of caveats and we would be better off designing a complete push-based metrics solution. See https://github.com/open-telemetry/opentelemetry-specification/issues/1078 and https://github.com/open-telemetry/oteps/pull/185

Author: Thanks for the feedback, I've updated the PR and mentioned "Support for Push mode is out of scope for the initial release."

Prometheus Exporter MUST support version `0.0.4` of the [Text-based
format](https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md#text-based-format).

Prometheus Exporter SHOULD only support [Cumulative
Temporality](../datamodel.md#temporality). Attempt to change the [Aggregation
Temporality](../datamodel.md#temporality) to

> [Delta](../datamodel.md#temporality) SHOULD be treated as an error. It is

Author: I've updated the wording, PTAL.

**Status**: [Experimental](../../document-status.md)

> Prometheus Exporter is a [Pull Metric Exporter](../sdk.md#pull-metric-exporter)

Reviewer: ```suggestion
A Prometheus Exporter is a [Pull Metric Exporter](../sdk.md#pull-metric-exporter)
```

which reports metrics by responding to the [Prometheus](https://prometheus.io/)
scraper requests.

> Prometheus Exporter MUST support [Pull mode](../sdk.md#pull-metric-exporter).

Reviewer: ```suggestion
A Prometheus Exporter MUST support [Pull mode](../sdk.md#pull-metric-exporter).
```

> Prometheus Exporter MUST NOT support [Push

Reviewer: ```suggestion
A Prometheus Exporter MUST NOT support [Push
```

mode](../sdk.md#push-metric-exporter).

> Prometheus Exporter MUST only support [Cumulative

Reviewer: ```suggestion
A Prometheus Exporter MUST only support [Cumulative
```

Temporality](../datamodel.md#temporality).

> Prometheus Exporter MUST support version `0.0.4` of the [Text-based

Reviewer: ```suggestion
A Prometheus Exporter MUST support version `0.0.4` of the [Text-based
```

format](https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md#text-based-format).

> Prometheus Exporter MAY support [OpenMetrics Text

Reviewer: ```suggestion
A Prometheus Exporter MAY support [OpenMetrics Text
```
